buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -804,6 +810,24 @@ private static void decodeString(CharsetDecoder decoder, ByteBuffer src, CharBuf
         }
     }
 
+    static String decodeString(ByteBuf src, int readerIndex, int len, Charset charset) {
```

Comment:
I noticed this function is frequently called with Charset of US_ASCII.   There is a deprecated (but faster) String constructor that takes Ascii bytes you might try.   (@louiscryan  called this "cheating" at the time)

Reply:
If the `String(char[],true)` constructor can be accessed via reflection (I know not possible in java9+ by default), then an additional copy could be also avoided in non-array-backed ascii case, i.e. perform the same logic as the deprecated ascii constructor but reading from the buffer directly.

Reply:
@carl-mastrangelo I think that is actual a good idea... let me do this.

Reply:
@carl-mastrangelo done

Reply:
would this also work for `ISO_8859_1` (single byte encoding)?

Reply:
@Scottmitch hmm not sure.... @carl-mastrangelo  ?

Reply:
The _variable_ name for the ctor is `ascii`  but the javadoc refers to the values as just "bytes".  Assuming you set `hibyte` correctly the implementation looks like it handles it.   (can't say for non-openjdk tho!)

Reply:
@trustin @Scottmitch @ejona86 @ejona86 hmm what you think about the change in behaviour ? @njhill thanks pointing out.

Reply:
I used this 'cheat' recently actually. I'm fine with that.

Reply:
I'm fine with cheating assuming we have a unit test which will fail.